### ISSUE TITLE
Add documentation for nothing

### ISSUE THREAD

**Issue by Christian24** [Polymer/lit-html#784]
It would be pretty cool if there was documentation for `nothing` sentinel value. Maybe it could go into https://github.com/Polymer/lit-html/blob/master/docs/_guide/05-template-reference.md or https://github.com/Polymer/lit-html/blob/master/docs/_guide/02-writing-templates.md with an example.

**Comment by KeithHenry** [comment]
How is `nothing` better than `null` or `undefined`?

Where should we `import` `nothing` from?

The proposed doc changes don't say.

**Comment by ruphin** [comment]
`nothing` has a very specific purpose. If you have some custom element `<my-element>` and that custom element has a slot with fallback content, the fallback content will not appear if you render any non-comment nodes as children of that element. (this includes empty TextNodes)

```js
// This will usually render an empty TextNode inside the element if content is falsy,
// so the slot fallback content will not appear
html`<my-element>${ content ? content : "" }</my-element>`
html`<my-element>${ content ? content : null }</my-element>`
html`<my-element>${ content ? content : undefined }</my-element>`

// This will remove all non-comment child nodes when content is falsy,
// so the slot fallback content will appear
html`<my-element>${ content ? content : nothing }</my-element>`
```

Other than this, it has no real purpose and you may as well use `""` (emptystring) or `undefined` or `null`.

`nothing` is a direct export of `lit-html.js`:

```js
import { nothing } from "https://unpkg.com/lit-html/lit-html.js?module";
```

**Comment by Christian24** [comment]
@KeithHenry Thanks for the feedback. I did leave out the import on purpose, because some examples seem to have them while others don't. I added the import now. I also made some efforts to explain what @ruphin has explained here, but probably not as well. 

@ruphin Thanks for the context. I do wonder though if it is a good best practice to use it though? I mean `nothing` could still be seen as a uniform way of rendering '', null or undefined.

**Comment by ruphin** [comment]
It really depends on a bunch of factors, like the context where it is being used. If it is being used in a switch style statement and the other part of the statement is a primitive like a String or a Number, then technically using `""` will have slightly better performance than `undefined` or `null`, which in turn have equal or slightly better performance than `nothing`. If the other part of the statement is a template or Node or some other complex object, then `nothing` may have slightly better performance than the others. Aditionally, `""`, `undefined`, and `null` may cause unexpected behaviours, but only in the above mentioned scenario.

Realistically we're talking in the order of microseconds, and from a performance perspective you can use any of these and not notice any difference. You can use whatever you prefer, and any 'best practice' is just opinionated.

**Comment by Christian24** [comment]
@ruphin Thanks for the explanation. So it basically boils down to string comparisons => `''`will be slightly faster, object comparison => `nothing` is slightly faster? The unexpected behaviors led my colleague and myself to stick to `nothing` in our project.

@KeithHenry I was wondering the same thing, but from reading the comments on the issue the team kinda expects `nothing` to be not so commonly used and therefore I guess it wouldn't be exported. I mean directives are also not exported and do require a separate installation of lit-html.

**Comment by justinfagnani** [comment]
This isn't true:
> I mean directives are also not exported and do require a separate installation of lit-html.

**Comment by ruphin** [comment]
It's not so much about the speed of comparisons, the effect on the DOM is different. If the current value is also a String and you change it to `""` (Or `null` or `undefined`), lit-html changes the TextContent of the TextNode to `""`. If the value changes to `nothing`, it will remove the TextNode from the DOM, and consequently have to insert a new TextNode once the value changes to a String again.

The difference can probably best be explained in once sentence: 

> `nothing` results in all nodes being removed from DOM entirely, and the other options result in a single TextNode with `""` as it's content.

This explains the effect it has on slotted content, and how it has subtle performance difference depending on what other content you're switching in and out.

**Comment by Christian24** [comment]
@ruphin Thanks, that does make a lot sense. If I find some time tomorrow, I will try to include that into the docs.


@justinfagnani 
> This isn't true:
> 
> > I mean directives are also not exported and do require a separate installation of lit-html.

How do you use directives in LitElement without installing lit-html along side? Maybe I am missing something trivial here, but I do not see them exported here: https://github.com/Polymer/lit-element/blob/master/src/lit-element.ts and I had asked about this on the LitElement repo a while back: https://github.com/Polymer/lit-element/issues/430

**Comment by ruphin** [comment]
If you install lit-element it will also install lit-html since it is a dependency. So if you have lit-element you have access to the directives through lit-html.

Adding lit-html as a dependency alongside lit-element may lead to having lit-html installed in two locations in node_modules, but they should be merged during the building process, as long as the versions are compatible.

**Comment by KeithHenry** [comment]
I'm a little confused by all this, but I think the purpose of `nothing` is to ensure that the content is skipped rather than an empty text nodes.

You want this in slots with fallback content.

But, wouldn't this make many updates _slower_? You want the empty node so that you're replacing it, rather than shuffling down the nodes after it?

That would be the behaviour with anything VDOM based (like React) that a lot people needing the documentation will be coming from.

So the only use case would be fallback for slots, in which case the name `nothing` describes what it _is_ but not what it's _for_. Something we should probably be clear about in the docs.

**Comment by ruphin** [comment]
It isn't necessarily slower. If the alternate content is not a primitive like a String or a Number, but instead another template, or array, or any other type of value, then `nothing` is actually more efficient because it does not create an empty TextNode.

Either way, the performance impact of these things is extremely minimal. If you are worried about this, you are probably spending energy to optimise the wrong things, and there are many other things you should be worried about that are much more impactful than this.

**Comment by KeithHenry** [comment]
> If you are worried about this...

_I'm_ not, but this is a new library with few established best practices. I think the documentation needs to be clear on this feature. Conditional rendering is something that users need to do a lot, and it's one of the first things every developer I've introduced to lit has asked me about (it's something I asked about #395), and responding with a lot of additional complexity can put them off.

I don't think it will make much difference to lit either way because of how templates work, but React/VDOM/JSX devs (which many new users reading the docs will be) will _expect_ it to.

We should be clear on the docs about what best practice is - I _think_ it's: use `nothing` for slots, otherwise `''`, `undefined` and `null` make no difference, use `''` as it's the fewest chars.

**Comment by Christian24** [comment]
@justinfagnani @ruphin Thanks, I had no idea. You learn something new everyday. 

@KeithHenry I do agree it should be documented, I might give working it into the nothing documentation a shot. However, I do think `undefined` and `null` are not really good options because of @ruphin's example here: https://github.com/Polymer/lit-html/issues/652#issuecomment-443253757 Just "quoting" here for simplicity:

````js
document.getElementById('container').innerHTML = null;
// <div id="container"></div>

document.getElementById('container').innerHTML = undefined;
// <div id="container">undefined</div>
````
So, basically the DOM is not consistent in how it deals with null or undefined => I would use either `''` or `nothing`.

**Comment by Christian24** [comment]
I did update my PR with more information on rendering no content (`''`, `null`, `undefined` and `nothing`). 
@ruphin I did use your example mentioned above, hope that is okay with you. I tried to put the performance aspects into my own words, not sure I succeeded. 

Hopefully it can be a start to get this documented as I agree with @KeithHenry it needs to go into the docs.

**Comment by arthurevans** [comment]
Fixed in #785.

**Comment by johnmellor** [comment]
It would also be nice to document how to check whether a return value is nothing. Presumably as it is a sentinel value you can compare if a value `=== nothing`?

**Comment by arthurevans** [comment]
@johnmellor I don't work on Lit anymore, but IIRC Lit shifted to rendering no node for any falsy value (empty string, null, `nothing`, etc.). So there's even less reason to use `nothing` that there was before (although there is another. use case for it when binding to an attribute--binding `nothing` removes the attribute).

https://lit.dev/docs/templates/expressions/#removing-child
https://lit.dev/docs/templates/expressions/#removing-attribute

You could of course compare a value to `nothing`. It's just a value. But I'm not aware of any API that returns `nothing`.

**Comment by arthurevans** [comment]
Also hi Lit fam :D
